hazelcast/src/main/java/com/hazelcast/scheduledexecutor/IScheduledExecutorService.java

```diff
@@ -0,0 +1,416 @@
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Distributed & durable implementation of {@link ScheduledExecutorService}.
+ * <code>IScheduledExecutorService</code> provides similar API to the <code>ScheduledExecutorService</code> with some
+ * exceptions but also additional methods like scheduling tasks on a specific member, on a member who is owner of a specific key,
+ * executing a tasks on multiple members etc.
+ *
+ *</p>
+ *
+ * Tasks (<tt>Runnable</tt> and/or <tt>Callable</tt>) scheduled on any partition through an <tt>IScheduledExecutorService</tt>,
+ * yield some durability characteristics.
+ * <ul>
+ *     <li>When a node goes down (up to <tt>durability</tt> config), the scheduled task will get re-scheduled
+ *     on a replica node.</li>
+ *     <li>In the event of a partition migration, the task will be re-scheduled on the destination node.</li>
+ * </ul>
+ * <b>Note: </b> The above characteristics don't apply when scheduled on a <tt>Member</tt>.
+ *
+ * <br/>
+ * <br/>
+ *
+ * Tasks that are holding state that needs to be also durable across partitions, will need to implement the {@link StatefulTask}
+ * interface.
+ *
+ * <br/>
+ * <br/>
+ *
+ * Upon scheduling a task acquires a resource handler, see {@link ScheduledTaskHandler}. The handler is generated before
+ * the actual scheduling of the task on the node, which allows for a way to access the future in an event of a node failure
+ * immediately after scheduling, and also guarantees no duplicates in the cluster by utilising a unique name per task.
+ * A name can also be defined by the user by having the <tt>Runnable</tt> or <tt>Callable</tt> implement the {@link NamedTask}.
+ * Alternatively, one can wrap any task using the {@link TaskUtils#named(String, Callable)} or
+ * {@link TaskUtils#named(String, Runnable)} for simplicity.
+ *
+ * <p/>
+ *
+ * One difference of this service in comparison to {@link ScheduledExecutorService} is the
+ * {@link #scheduleAtFixedRate(Runnable, long, long, TimeUnit)} which has similar semantic
+ * to {@link java.util.concurrent.ScheduledExecutorService#scheduleAtFixedRate(Runnable, long, long, TimeUnit)}. It
+ * guarantees a task won't be executed by multiple threads concurrently. The difference is that this service will
+ * skip a scheduled execution if another thread is still running the same task, instead of postponing its execution.
+ * To emphasize this difference the method is called <code>scheduleAtFixedRate</code>
+ * instead of <code>scheduleAtFixedRate</code>
+ * <br/><br/>
+ * The other difference is this service does not offer an equivalent of
+ * {@link java.util.concurrent.ScheduledExecutorService#scheduleWithFixedDelay(Runnable, long, long, TimeUnit)}
+ */
+@Beta
+public interface IScheduledExecutorService extends DistributedObject {
```

Comment:
what was the reason that we didn't extend `ScheduledExecutorService`? In other distributed objects if there is an interface available we are extending it, putting some examples here:

`public interface ILock extends Lock, DistributedObject`
`public interface IMap<K, V> extends ConcurrentMap`
`public interface ICondition extends Condition `
`public interface ICollection<E> extends Collection<E>, DistributedObject`